File: March-2023/C++/SymmetricTree.cxx
// 101. Symmetric Tree
// just make sure the recrusion for left and right comparison both return true using &&
class Solution {
public:
    bool isTrue(TreeNode* left, TreeNode* right){
        if(left == nullptr && right == nullptr) return true;
        if(left == nullptr || right == nullptr) return false;
        if(left->val != right->val) return false;
        return isTrue(left->left, right->right) && isTrue(left->right, right->left);
    }
    
    bool isSymmetric(TreeNode* root) {
        // for single node
        if(root == nullptr) return true;
        // if anyone is null, that means its not a mirror
        if(root->left == nullptr && root->right == nullptr) return true;
        if(root->left == nullptr || root->right == nullptr) return false;
        return isTrue(root->left, root->right);
    }
};

File: March-2023/Java/SymmetricTree.java
// 101. Symmetric Tree
// just make sure the recrusion for left and right comparison both return true using &&
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isTrue(TreeNode left, TreeNode right){
        if(left==null && right==null) return true;
        if(left==null || right==null) return false;        
        if(left.val!=right.val) return false;       
        return isTrue(left.left, right.right) && isTrue(left.right,right.left);       
    }
     
    public boolean isSymmetric(TreeNode root) {
        // if only one node is present
        if(root.left==null && root.right==null) return true;
      // if any one is null its not a mirror
           if(root.left==null || root.right==null) return false;
        if(isTrue(root.left,root.right)){
            return true;
        }
        return false;
    }
}
